Re-include RBF replacement txs in fee estimation

darosior opened the pull request:
This effectively reverts #9519.

RBF is now largely in use on the network (signaled for by around 20% of
all transactions on average) and replacement logic is implemented in
most end-user wallets. The rate of replaced transactions is also
expected to rise as fee-bumping techniques are being developed for
pre-signed transaction ("L2") protocols.

darosior committed:
Re-include RBF replacement txs in fee estimation

This effectively reverts de1ae324bf3fb7451c1008a1a9721ff9f469533b.

RBF is now largely in use on the network (signaled for by around 20% of
all transactions on average) and replacement logic is implemented in
most end-user wallets. The rate of replaced transactions is also
expected to rise as fee-bumping techniques are being developed for
pre-signed transaction ("L2") protocols.

Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

MarcoFalke commented on src/validation.cpp:
```diff
@@ -982,7 +982,6 @@ bool MemPoolAccept::Finalize(const ATMPArgs& args, Workspace& ws)
     const CAmount& nModifiedFees = ws.m_modified_fees;
     const CAmount& nConflictingFees = ws.m_conflicting_fees;
     const size_t& nConflictingSize = ws.m_conflicting_size;
-    const bool fReplacementTransaction = ws.m_replacement_transaction;
```
Can also remove `m_replacement_transaction`, now that it is unused and inline `setConflicts.size()` again?

darosior replied:
Done.

prayank23 commented:
Concept ACK

darosior commented:
Asking the mempool statistics reference on IRC, @0xB10C pointed out to me there are between 2000 and 6000 replacements per day at the moment.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -805,8 +803,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
     // mempool consistency for us.
-    fReplacementTransaction = setConflicts.size();
-    if (fReplacementTransaction)
+    if (setConflicts.size())
     {
```
nit: braces go on the same line

```suggestion
    if (setConflicts.size()) {
```

darosior replied:
Done.

maflcko reviewed (commented):
Out of interest, did anyone collect stats on how many txs are replacements? Edit: see previous comment

prayank23 commented:
> Out of interest, did anyone collect stats on how many txs are replacements? Edit: see previous comment

Although 0xb10c shared there are 2000-6000 replacement tx per day , I don't think it matters for few reasons:

1. Fee estimation is a meme IMO or maybe helps in UX for newbies. You cannot predict if a fee rate will confirm the transaction in next N blocks. Context: https://bitcoin.stackexchange.com/questions/105860/what-are-we-trying-to-predict-in-fee-estimation-and-why/106129
2. If fee estimates are used, they should be available for all type of transactions and not based on % of txs per day. Rationale in PR https://github.com/bitcoin/bitcoin/pull/9519 was not good enough to change something. Not sure about the ACKs in that PR.
3. If any % still matters to document in PR, it should be RBF per day because if a user did transaction with RBF enabled, fee estimate may be required by replacement tx. Total transaction per day around 200k so 20% is 40k RBF transactions. We shouldn't wait for usage or other node implementations/wallets to do better before doing things that make sense.

0xB10C commented:
> Out of interest, did anyone collect stats on how many txs are replacements? Edit: see previous comment

![image](https://user-images.githubusercontent.com/19157360/126814167-15aae792-cc2b-46a5-8445-bd37fdfca1a2.png)

josedanielrojaspixel commented:
Llevo el timon que es lo que inporta nadieme inpresionaEnviado desde mi HUAWEI Mate 20 Pro-------- Mensaje original --------De: Prayank ***@***.***>Fecha: vie., 23 jul. 2021 12:36 p. m.Para: bitcoin/bitcoin ***@***.***>CC: Subscribed ***@***.***>Asunto: Re: [bitcoin/bitcoin] Re-include RBF replacement txs in fee estimation (#22539)

Out of interest, did anyone collect stats on how many txs are replacements? Edit: see previous comment

Although 0xb10c shared there are 2000-6000 replacement tx per day , I don't think it matters for few reasons:
Fee estimation is a meme IMO or maybe helps in UX for newbies. You cannot predict if a fee rate will confirm the transaction in next N blocks. Context: https://bitcoin.stackexchange.com/questions/105860/what-are-we-trying-to-predict-in-fee-estimation-and-why/106129If fee estimates are used, they should be available for all type of transactions and not based on % of txs per day. Rationale in PR #9519 was not good enough to change something. Not sure about the ACKs in that PR.If any % still matters to document in PR, it should be RBF per day because if a user did transaction with RBF enabled, fee estimate may be required by replacement tx. Total transaction per day around 200k so 20% is 40k RBF transactions. We shouldn't wait for usage or other node implementations/wallets to do better before doing things that make sense.

—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or unsubscribe.

Rspigler commented:
Concept ACK, due to reasons stated in OP

benthecarman reviewed (approved):
utACK 4abe6aab798fcb9d80228d4e49844b96690bfa2f

darosior commented:
For what it's worth, I've been comparing running this on 2 pretty similar nodes for the past week. This didn't affect estimates, it must be because we are currently in a pretty low fee period.

tryphe reviewed (approved):
untested ACK 4abe6aab798fcb9d80228d4e49844b96690bfa2f (edited hash, sorry clipboard was bugged)

Concept is simple and code changes are very straightforward (only one line is effectively changed).

theStack commented on src/validation.cpp:
```diff
@@ -805,9 +803,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
-    fReplacementTransaction = setConflicts.size();
-    if (fReplacementTransaction)
-    {
+    if (setConflicts.size()) {
```
nit: Not sure if we have this in the coding guidelines, but personally I'm not a fan of implicit int-to-bool casts. Could avoid it via e.g.
```suggestion
    if (setConflicts.size() > 0) {
```
or
```suggestion
    if (!setConflicts.empty()) {
```

theStack reviewed (approved):
Concept and code review ACK 4abe6aab798fcb9d80228d4e49844b96690bfa2f

josibake reviewed (approved):
utACK https://github.com/bitcoin/bitcoin/pull/22539/commits/4abe6aab798fcb9d80228d4e49844b96690bfa2f

reading through #9519 , the general consensus seemed to be "this will be trivial to and should be reverted once RBF is widely adopted by miners." 

i didn't get the sense anyone was arguing to never include RBF in fee estimation, so given that RBF is widely adopted, this change makes sense.

Zero-1729 reviewed (approved):
crACK 4abe6aab798fcb9d80228d4e49844b96690bfa2f

Change makes sense given the wide adoption of RBF txs. Plus, as stated above, the original PR that excluded RBFs did state a potential revert in the future once there's increased use in the wild (when widely adopted as "miner policy").

**Note**: I am leaning towards the code suggestion [here](https://github.com/bitcoin/bitcoin/pull/22539#discussion_r680198364), specifically `if (setConflicts.size() > 0) {`, to make the condition more explicit about the expected value of `setConflicts.size()` (as was similarly done [here](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L712) in the same file). It's a non-blocking nit, so feel free to ignore.

glozow reviewed (commented):
code review ACK 4abe6aab798fcb9d80228d4e49844b96690bfa2f

If we're at the point where most of the network and miners support RBF,  this makes our fee estimates more accurate. My concern correctness-wise was making sure that only the new transaction is considered in fee estimation (I don't have much knowledge of `CBlockPolicyEstimator`). I've checked that when `Finalize()` calls `RemoveStaged()` which calls `removeUnchecked()` for each transaction, the mempool will call `minerPolicyEstimator->removeTx()` for the removed transactions.

darosior commented:
Rebased on master after #22796 merge and addressed @theStack 's nit.

theStack reviewed (approved):
re-ACK 5b29258076a3665100400768de8124be64eb4c6f 🌴 

Verified via `git range-diff 4abe6aab...5b292580` that changes since my last ACK are only rebase-related, plus the slight improvement in the if condition (avoiding implicit int-to-bool cast).

josibake commented:
re-ACK https://github.com/bitcoin/bitcoin/pull/22539/commits/5b29258076a3665100400768de8124be64eb4c6f

Zero-1729 reviewed (approved):
re-ACK 5b29258 🧉

fanquake commented:
@darosior sorry for making you rebase again, but this change is more straightforward than some other PRs in that regard.

darosior commented:
Rebased

> sorry for making you rebase again, but this change is more straightforward

Sure

glozow reviewed (commented):
reACK fab8ce071566033316189283684340424cb2f7d0. Nice cleanup as well.

theStack reviewed (approved):
re-ACK fab8ce071566033316189283684340424cb2f7d0

Verified via `git range-diff 5b292580...fab8ce07` that changes since my last ACK are only rebase-related.

benthecarman reviewed (approved):
reACK fab8ce071566033316189283684340424cb2f7d0

prayank23 commented:
ACK https://github.com/bitcoin/bitcoin/pull/22539/commits/fab8ce071566033316189283684340424cb2f7d0

Although I still have the same opinion: https://github.com/bitcoin/bitcoin/pull/22539#issuecomment-885759874

Lets not restrict it for some transactions because of numbers.

Zero-1729 reviewed (approved):
re-ACK fab8ce071566033316189283684340424cb2f7d0

Clean rebase.

darosior committed:
qa: split run_test into smaller parts

Let's not have run_test get into a giant function as we add more tests

Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

darosior committed:
qa: test fee estimation with replacement transactions

Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

darosior committed:
Add release notes for fee est with replacement txs

Signed-off-by: Antoine Poinsot <darosior@protonmail.com>

darosior commented:
Thanks everyone for following, please bear with me one more time :)

Rebased on master to solve conflict, added a functional test demonstrating behaviour (similarly to #23074 ) and added release notes.

darosior commented:
So i've been plotting the estimates at each block on both a node with this patch and one without. Here are the results for the past 3 days (still running)

![image](https://user-images.githubusercontent.com/22457751/134709338-fd2ac969-7c41-4df8-9599-377b8d456552.png)

Feerate is low at the moment so there is hardly any divergence.
Data for generating this plot can be found here https://github.com/darosior/fee_estimation_hacks

darosior commented:
Rebased following conflict with #22722 .

Here is the plot of the estimates with and without this patch for the past 8 days:
![image](https://user-images.githubusercontent.com/22457751/135310699-0c2a2737-246b-4811-bd5d-0035b11950f9.png)

prayank23 commented:
reACK https://github.com/bitcoin/bitcoin/pull/22539/commits/3b613722f6b895d7b268b3f878fddfc888381226

Zero-1729 reviewed (commented):
re-ACK 3b613722f6b895d7b268b3f878fddfc888381226

benthecarman reviewed (approved):
reACK 3b613722f6b895d7b268b3f878fddfc888381226

glozow commented on doc/release-notes-22539.md:
```diff
@@ -0,0 +1,8 @@
+P2P and network changes
+-----------------------
+
+- Fee estimation now takes the feerate of replacement (RBF) transactions into
```
I have a question about the desired behavior because it's a subtle thing that isn't mentioned here: with this PR, if a transaction is RBF'ed in our mempool, we add the replacement transaction and remove the original transaction(s) from the fee estimator. If the original transactions end up being mined in a block anyway (e.g. a miner had already started working on it before the replacement transaction was accepted), we lose that data in fee estimation. Is this what we want? If yes, maybe document?

darosior replied:
> Is this what we want?

Yes, because there is nothing we can do about that. And such a race sounds to be very improbable.. (are there still mining pools refusing RBFs for whatever reason?)

>  If yes, maybe document?

This is not a change in behaviour, i don't think it should be in the changelog. Before this patch if you accept a replacement transaction in your mempool but the original tx ends up being mined you would lose the data anyways.

glozow replied:
Ah you're right, we're already removing the original transactions. My mistake!

glozow reviewed (commented):
Thanks for adding a test! Sorry it's taken me some time to review - also ended up with an additional question.

Other possible test extensions:
- check that children of the original transactions aren't included in fee estimation (all evicted)
- transaction is RBF'ed in our mempool, but the original transaction is included in a block
- same-txid-different-witness transaction i.e. different feerate is included in a block

darosior commented on test/functional/feature_fee_estimation.py:
```diff
@@ -279,11 +283,100 @@ def run_test(self):
+                txid = send_tx(node, u, low_feerate)
+                utxos_to_respend.append(u)
+                txids_to_replace.append(txid)
+            # Broadcast 5 low fee transaction which don't need to
+            for _ in range(5):
+                send_tx(node, utxos.pop(0), low_feerate)
```
> I'm a little confused about why there needs to be 10% of the transactions confirming with low feerate?

There does not need to, i just wanted to simulate a more realistic scenario than just triggering the specific behaviour.

darosior commented on test/functional/feature_fee_estimation.py:
```diff
@@ -279,11 +283,100 @@ def run_test(self):
+            self.sync_mempools(wait=.1, nodes=[node, miner])
+            for txid in txids_to_replace:
+                miner.prioritisetransaction(txid=txid, fee_delta=-COIN)
+            self.generate(miner, 1)
```
Yes it is. The point is to simulate a somewhat real life scenario "broadcast a low fee tx, it does not confirm, it gets RBF'd, the replacement gets confirmed". You would not RBF before a single block was mined

glozow replied:
ah ok, putting it that way makes sense

glozow commented on test/functional/feature_fee_estimation.py:
```diff
@@ -155,6 +156,21 @@ def check_estimates(node, fees_seen):
     check_smart_estimates(node, fees_seen)
 
+
+def send_tx(node, utxo, feerate):
+    """Broadcast a 1in-1out transaction with a specific input and feerate (sat/vb)."""
+    overhead, op, scriptsig, nseq, value, spk = 10, 36, 5, 4, 8, 24
+    tx_size = overhead + op + scriptsig + nseq + value + spk
+    fee = tx_size * feerate
+
+    tx = CTransaction()
+    tx.vin = [CTxIn(COutPoint(int(utxo["txid"], 16), utxo["vout"]), SCRIPT_SIG[utxo["vout"]])]
+    tx.vout = [CTxOut(int(utxo["amount"] * COIN) - fee, P2SH_1)]
+    txid = node.sendrawtransaction(tx.serialize().hex())
+
+    return txid
```
Ah I hadn't seen that - great!

darosior commented:
> transaction is RBF'ed in our mempool, but the original transaction is included in a block

That's a very good point. I think with this patch we would keep the replacement tx in the buckets until it expires..

glozow commented:
> I think with this patch we would keep the replacement tx in the buckets until it expires..

Actually, I think it would be removed [here](https://github.com/bitcoin/bitcoin/blob/master/src/txmempool.cpp#L639) in `removeForBlock()` since it's conflicted out of the mempool. So that's fine, we'd just lose a data point. I ask because in some cases we keep the original transaction (e.g. we'll put it in `vExtraTxnForCompact` for reconstructing compact blocks), so I wondered if that's something we want here.

glozow reviewed (commented):
ACK 3b613722f6b895d7b268b3f878fddfc888381226

Thanks for patiently answering my questions! Verified that 053415b297b8665f2d2c4dce7c2c54bcc5298ef4 is move-only, and the added functional test looks correct to me. Will take my test improvement-related comments to #23075.

darosior commented:
> Actually, I think it would be removed here when removeForBlock() calls removeConflicts here since it's conflicted out of the mempool. So that's fine, we'd just lose a data point.

So at first i did not think it was fine. We in this case call `removeTx` with `inBlock` set to `false`, ie counting this removal as a failure to confirm. This could have lead to wrecked estimations because we would count a higher fee tx would be counted as failed.

But actually, we'd only count it as a failure if it didn't confirm for at least 12 blocks (for short stats, 24 and 42 for med and long). It's plenty of time for the replacement to have propagated and makes this type of race so unlikely that i think it's fine.

Thank you for raising this point!

darosior commented:
@tryphe @theStack @josibake do you mind re-[reviewing-and-]ACKing? I think we are getting close to merge and no other conflicting PR reported by Drahbot seem closer to merge.

theStack commented on test/functional/feature_fee_estimation.py:
```diff
@@ -155,6 +156,21 @@ def check_estimates(node, fees_seen):
+
+def send_tx(node, utxo, feerate):
+    """Broadcast a 1in-1out transaction with a specific input and feerate (sat/vb)."""
+    overhead, op, scriptsig, nseq, value, spk = 10, 36, 5, 4, 8, 24
+    tx_size = overhead + op + scriptsig + nseq + value + spk
+    fee = tx_size * feerate
+
+    tx = CTransaction()
+    tx.vin = [CTxIn(COutPoint(int(utxo["txid"], 16), utxo["vout"]), SCRIPT_SIG[utxo["vout"]])]
+    tx.vout = [CTxOut(int(utxo["amount"] * COIN) - fee, P2SH_1)]
```
To avoid manual tx size calculation with magic numbers, you could simply create the `CTransaction` first (without the fee), determine tx size via `len(tx.serialize())` and then subtract the fee. 

```suggestion
    tx = CTransaction()
    tx.vin = [CTxIn(COutPoint(int(utxo["txid"], 16), utxo["vout"]), SCRIPT_SIG[utxo["vout"]])]
    tx.vout = [CTxOut(int(utxo["amount"] * COIN), P2SH_1)]
    fee = len(tx.serialize()) * feerate
    tx.vout[0].nValue -= fee
```

darosior replied:
Oh right since we use legacy txs in this test it's indeed much cleaner. However i'd like to not invalidate 4 ACKs on a cosmetic change. Do you mind if i make it a follow-up in #23075 ?

theStack commented on test/functional/feature_fee_estimation.py:
```diff
@@ -271,6 +287,60 @@ def test_feerate_mempoolminfee(self):
+            self.generate(miner, 1)
+            self.sync_blocks(wait=.1, nodes=[node, miner])
+            # RBF the low-fee transactions
+            while True:
+                try:
+                    u = utxos_to_respend.pop(0)
+                    send_tx(node, u, high_feerate)
+                except IndexError:
+                    break
```
Might be a bit shorter and more pythonic without exception:
```suggestion
            while utxos_to_respond:
                u = utxos_to_respend.pop(0)
                send_tx(node, u, high_feerate)
```

darosior replied:
Indeed :man_facepalming: but same rationale as above, as ugly as it is i'd rather address it in #23075

theStack reviewed (approved):
re-ACK 3b613722f6b895d7b268b3f878fddfc888381226 🍪 

Verified that since my last ACK a rebase was done and a test was added which LGTM. Being extra-paranoid, I also checked that the test fails on the master branch on the estimated fee assertion at the end of `sanity_check_rbf_estimates`.

@darosior: I agree that the proposed fixups can be done in a follow-up, was not intending to block a merge (I kind of missed reading previous comments about fixups and PR #23075 that is already prepared for that purpose). BTW I also support the idea of using MiniWallet for creating txs, as suggested by gloria.

darosior commented:
For what it's worth here is again another diagram for last ~2 wee
![702743_705101](https://user-images.githubusercontent.com/22457751/137485633-9867e8ec-dbf1-48a1-969c-9354e8e7399b.png)
ks
